metrics/perplexity/perplexity.py

```diff
@@ -188,3 +185,46 @@ def _compute(self, predictions, model_id, batch_size: int = 16, add_start_token:
             ppls += perplexity_batch.tolist()
 
         return {"perplexities": ppls, "mean_perplexity": np.mean(ppls)}
+
+    def _compute(self, logits=None, labels=None, attention_mask=None):
```

Comment:
The convention is to use `predictions` (=`logits`) and `references`(=`labels`), unless it makes no sense at all, e.g. in RL. Not that these should be args and their type should be specified in `features`.

Reply:
Note that the attention mask and the use of tensors in general breaks a bit the pattern here:
- in general metrics take data with either `add` (single sample), `add_batch` (a batch), `compute` (all data at once or the data previously added through `add`/`add_batch`). The data is stored with Arrow and when loaded for `_compute` the data is passed as a Python list of samples.

As far as I can tell the `attention_mask` is used because the logits/labels might sometimes be shorter than the batch's global size. I think in many cases the model is evaluated on equal sized chunks so this wouldn't be necessary. So we can either assume it is not necessary or that the user slices each sample to it's actual sequence length and adds it as a list. In both cases we wouldn't need the attention mask, right?

Also we'll get the data as lists/numpy arrays so it might be easier to do the computations in numpy. But if it's too annoying to do e.g. `CrossEntropyLoss` in pure numpy we can also keep `torch` as a dependency.

Reply:
Oh yes, you can totally input a tensor. What I meant is that when `_compute` is called the data is loaded as a list of samples, not a single tensor/array but rather a list of them. So you can either assume that the user fed tensors of equal sequence length and just stack them or treat them as tensors of unequal sequence length and compute ppl for each sample and aggregate them. If we go for the latter approach we also wouldn't need the attention mask.

Note that features currently can't be optional (we could add that as a feature though) and so we can't pass the attention mask as a data feature but must use the compute kwargs. That means that it works a bit different with `add`/`add_batch`/`compute`:

```python
metric = load("perplexity")

# with `compute` alone:
metric.compute(predictions=predictions, references=references, attention_mask=attention_mask)

# for same with `add`
for p, r in zip(predictions, references):
    metric.add(metric.add(prediction=p, reference=r))
metric.compute(attention_mask=attention_mask)
```

Also this wouldn't work well in a distributed setting I think with the way `add` and `compute` are setup. Not sure what the best approach is in this case.